lib/src/characters.dart

```diff
@@ -292,85 +195,355 @@ abstract class Characters implements Iterable<String> {
   String toString();
 }
 
-/// Iterator over characters of a string.
+/// A sub-sequence of a [Characters].
+///
+/// The sub-sequence is a range of consecutive characters in [source],
+/// corresponding to a start and end position in the source sequence.
+/// The range may even be empty, but will still correspond to a position,
+/// only one where start and end is the same position.
+///
+/// The source sequence can be separated into the *previous* characters,
+/// those before the range, the range itself, and the *next* characters,
+/// those after the range.
+///
+/// Inside the range, operations may look at the *first* characters of the
+/// range, those right after the range start,
+/// or the *last* characters of the range, those right before the range end.
 ///
-/// Characters are Unicode grapheme clusters represented as substrings
-/// of the original string.
+/// The range of a [CharacterRange] can be updated to include more
+/// characters at either end, or to drop characters from either end,
+/// or to simply move to a completely new range.
 ///
-/// The [start] and [end] indices will iterate the grapheme cluster
-/// boundaries of the string while the [Character] is iterating the
-/// grapheme clusters. A string with *n* grapheme clusters will have
-/// *n + 1* boundaries (except when *n* is zero, then there are also
-/// zero boundaries). Those boundaries can be accessed as, for example:
-/// ```dart
-/// Iterable<int> graphemeClusterBoundaries(String string) sync* {
-///   if (string.isEmpty) return;
-///   var char = Characters(string).iterator;
-///   var hasNext = false;
-///   do {
-///     hasNext = char.moveNext();
-///     yield char.start;
-///   } while (hasNext);
-/// }
-/// ```
-abstract class Character implements BidirectionalIterator<String> {
+/// The character range implements [Iterator]
+/// (actually an [BidirectionalIterator] of [String]).
+/// The [moveNext] operation, when called with no argument,
+/// iterates the *next* single characters of the [source] sequence.
```

Comment:
Overall, I like the set of operations here and I mostly like the names but some
of them get pretty confusing, like `includeAfterPrevious()` and
`nextStartsWith()`. I tried to come up with a few changes and apply them
systematically like you did:

### Entities

*   **"next"** – A CharacterRange following this one, in the sense meant by
    `Iterator.moveNext()`.
*   **"previous"** – A CharacterRange preceding this one, in the sense meant by
    `BidirectionalIterator.movePrevious()`.
*   **"preceding"** – The remainder of string before this range.
*   **"following"** – The remainder of string after this range.
*   **"first"** - The leftmost occurence of some target Characters.
*   **"last"** - The rightmost occurence of some target Characters.
*   **"start"** – The leftmost end of the range.
*   **"end"** – The rightmost end of the range.

### Operations

*   **"move"** – Move to a new range adjacent to this one. Return a bool to
    indicate success.
*   **"collapse"** – Change the range to be just the referred to entity.
*   **"include"** – Add the preceding or following remainders to the range.
*   **"expand"** – Keep the current characters in the range but potentially grow
    the range to include more characters.
*   **"drop"** – Remove characters from the range.

### Modifiers:

*   **"to"** – Identifies the target entity sought by the operation. The target
    will either be a word following this, or the method's argument if `to` is
    the last word in the name. Unless specified otherwise, searching proceeds
    forwards. Omitted if the target is obviously the method argument.

*   **"back"** or **"backwards"** – Indicates that searching goes backwards. Use
    "backwards" when "back" could ambigiously refer to the back of something.
    This indicates *direction*, not *destination*.

*   **"until"** – Do not include the target entity itself in the resulting
    range. By default, the sought entity is included.

*   **"while"** – The target is the consecutive series of characters that match
    a given predicate.

That leads to these changes:

```dart
abstract class CharacterRange implements BidirectionalIterator<String> {
  /// Makes the range be next occurrence of [target] after the current range.
  bool moveNext([Characters target]); // -> moveNext(), moveTo(target)
  /// Makes the range be the characters up to the next occurrence of [target].
  bool moveUntilNext(Characters target); // -> moveUntil()
  /// Makes the range be previous occurrence of [target] before the current range.
  bool movePrevious([Characters target]); // -> movePrevious(), moveBackTo(target)
  /// Makes the range be the characters after to the previous occurrence of [target].
  bool moveAfterPrevious(Characters target); // -> moveBackUntil()
  /// Finds the first instance of [target] inside the current range.
  bool moveFirst([Characters target]); // -> collapseToFirst()
  /// Finds the last instance of [target] inside the current range.
  bool moveLast([Characters target]); // -> collapseToLast()

  /// Collapses the range to its start.
  void collapseStart(); // -> collapseToStart()
  /// Collapses the range to its end.
  void collapseEnd(); // -> collapseToEnd()

  /// Starts the range at the start of the source characters.
  void includeAllPrevious(); // -> includePreceding()
  /// Ends the range at the end of the source characters.
  void includeAllNext(); // -> includeFollowing()

  /// Includes the next occurrence of [target] into the current range.
  bool includeNext([Characters target]); // -> expandTo()
  /// Includes characters until the next occurrence of [target] into the current range.
  bool includeUntilNext(Characters target); // -> expandUntil()
  /// Includes the next single characters while they are accepted by [test].
  void includeNextWhile(bool test(String character)); // -> expandWhile()
  /// Includes the previous occurrence of [target] into the current range.
  bool includePrevious([Characters target]); // -> expandBackwards()
  /// Includes characters after the previous occurrence of [target].
  bool includeAfterPrevious(Characters target); // -> expandBackwardsUntil()
  /// Includes the previous single characters while they are accepted by [test].
  void includePreviousWhile(bool test(String character)); // -> expandBackwardsWhile()

  /// Drops the first occurrence of [target] from the current range.
  bool dropFirst([Characters target]); // dropTo()
  /// Drops characters before the first occurrence of [target].
  bool dropUntilFirst(Characters target); // dropUntil()
  /// Drops characters from the last occurrence of [target].
  bool dropLast([Characters target]); // dropBackTo()
  /// Drops characters afer the last occurrence of [target] from the current range.
  bool dropAfterLast(Characters target); // dropBackUntil()
  /// Drops initial single characters accepted by [test].
  void dropFirstWhile(bool test(String characters)); // dropWhile()
  /// Drops final single characters accepted by [test].
  void dropLastWhile(bool test(String characters)); // dropBackWhile()

  /// Whether the current range starts with [characters].
  bool startsWith(Characters characters);
  /// Whether the current range ends with [characters].
  bool endsWith(Characters characters);

  /// Whether the current range is preceeded by [characters].
  bool nextStartsWith(Characters characters); // precedingEndsWith()
  /// Whether the current range is followed by [characters].
  bool previousEndsWith(Characters characters); // followingStartsWith()
}
```

What do you think?

Reply:
Very systematic! I like it!

I did try to avoid "preceding" and "following". Maybe that wasn't a good idea.

The `includeNext`/`includePrevious` members have has two options. Should they be `expandToNext()` and `expandTo(something)`? (ditto previous)